File: backend/src/services/deliveryService.js
import * as orderRepo from '../repositories/orderRepository.js';
import { findUserByUserId } from '../repositories/userRepository.js';

const TRACKABLE = ['shipped', 'delivered'];

/**
 * The seller (or their courier) pushes a position while an order is in transit.
 * Only the seller who owns the order may write it.
 */
export const updateLocation = async (orderId, actor, { lat, lng, etaMinutes, courierName, courierContact }) => {
  const order = await orderRepo.findOrderById(orderId);
  if (!order) throw new Error('Order not found');

  // Two people may say where this order is: the seller who owns it, and the
  // rider actually carrying it. Assignment is what grants the rider that -
  // working for the seller is not enough on its own.
  const userId = typeof actor === 'object' ? actor.userId : actor;
  const isSeller = order.sellerId === userId;
  const isAssignedRider = order.delivery?.riderUserId === userId;

  if (!isSeller && !isAssignedRider) throw new Error('Not your order');

  if (!TRACKABLE.includes(order.status)) {
    throw new Error('Tracking only runs once the order is shipped.');
  }

  const latN = Number(lat);
  const lngN = Number(lng);
  if (!Number.isFinite(latN) || latN < -90 || latN > 90) throw new Error('Invalid latitude');
  if (!Number.isFinite(lngN) || lngN < -180 || lngN > 180) throw new Error('Invalid longitude');

  order.delivery = {
    ...(order.delivery?.toObject?.() || order.delivery || {}),
    lat: latN,
    lng: lngN,
    etaMinutes: etaMinutes != null ? Number(etaMinutes) : order.delivery?.etaMinutes ?? null,
    courierName: courierName ?? order.delivery?.courierName ?? '',
    courierContact: courierContact ?? order.delivery?.courierContact ?? '',
    updatedAt: new Date(),
    isLive: true,
  };

  await order.save();
  return order.delivery;
};

/** Both the buyer who placed the order and the seller may read the position. */
export const getTracking = async (orderId, user) => {
  const order = await orderRepo.findOrderById(orderId);
  if (!order) throw new Error('Order not found');

  const isSeller = order.sellerId === user.userId;
  const isBuyer = order.buyerUserId === user.userId;
  const isRider = order.delivery?.riderUserId === user.userId;
  if (!isSeller && !isBuyer && !isRider && user.role !== 'superadmin') {
    throw new Error('This order is not yours');
  }

  const d = order.delivery || {};
  const r = order.route || {};

  // The snapshot is the truth once it exists. When it does not - an order
  // placed before the seller had pinned their shop - their current pickup
  // point stands in, so setting it late fixes the orders already out there
  // rather than only the next ones.
  let pickupLat = r.pickupLat ?? null;
  let pickupLng = r.pickupLng ?? null;
  let pickupAddress = r.pickupAddress || '';

  if (pickupLat == null || pickupLng == null) {
    const seller = await findUserByUserId(order.sellerId);
    pickupLat = seller?.pickupLat ?? null;
    pickupLng = seller?.pickupLng ?? null;
    pickupAddress = pickupAddress || seller?.pickupAddress || '';
  }

  return {
    orderNumber: order.orderNumber,
    status: order.status,
    productName: order.productName || '',
    isLive: Boolean(d.isLive && TRACKABLE.includes(order.status)),
    lat: d.lat ?? null,
    lng: d.lng ?? null,
    etaMinutes: d.etaMinutes ?? null,
    courierName: d.courierName || '',
    courierContact: d.courierContact || '',
    updatedAt: d.updatedAt || null,
    destination: order.deliveryAddress || '',

    /** The assigned rider, once the seller has put somebody on it. */
    riderUserId: d.riderUserId ?? null,

    // The photo taken at the door, and when. Behind /api/files - the buyer's
    // gate is in it - and readable by the three people on the delivery.
    proofOfDelivery: d.proofOfDelivery || '',
    deliveredAt: d.deliveredAt || null,
    deliveryNote: d.deliveryNote || '',

    // The two fixed ends. Both are known the moment an order exists, so the
    // map has something to show long before anybody shares a position.
    pickup: {
      lat: pickupLat,
      lng: pickupLng,
      address: pickupAddress,
    },
    dropoff: {
      lat: r.dropoffLat ?? null,
      lng: r.dropoffLng ?? null,
      address: order.deliveryAddress || '',
      // How good that point is. A barangay centre must never be presented as
      // somebody's front door.
      precision: r.dropoffPrecision || '',
    },
  };
};

/** Called when an order completes: stop broadcasting a rider's position. */
export const stopTracking = async (order) => {
  if (!order?.delivery) return;
  order.delivery.isLive = false;
  order.delivery.lat = null;
  order.delivery.lng = null;
  await order.save();
};

File: mobile/lib/services/delivery_service.dart
import 'api_client.dart';

/// A point on the map, with the words that go under it.
class DeliveryPoint {
  const DeliveryPoint({
    this.lat,
    this.lng,
    this.address = '',
    this.precision = '',
  });

  final double? lat;
  final double? lng;
  final String address;

  /// `exact` when the buyer pinned their own door; `approximate` when the
  /// point is the barangay the address names. Never shown as the same thing -
  /// a rider sent to a barangay centre believing it is a house has been
  /// misled by the map.
  final String precision;

  bool get isApproximate => precision == 'approximate';

  /// Whether this end can actually be drawn.
  ///
  /// An address that was typed but never pinned has words and no point - the
  /// rider can still read it, but the map has nothing to mark.
  bool get hasPoint => lat != null && lng != null;

  factory DeliveryPoint.fromJson(Map<String, dynamic>? json) {
    if (json == null) return const DeliveryPoint();

    return DeliveryPoint(
      lat: (json['lat'] as num?)?.toDouble(),
      lng: (json['lng'] as num?)?.toDouble(),
      address: (json['address'] ?? '').toString(),
      precision: (json['precision'] ?? '').toString(),
    );
  }
}

/// Where one order line is, between the shop it came from and the door it is
/// going to.
class DeliveryTracking {
  const DeliveryTracking({
    required this.orderNumber,
    required this.status,
    required this.productName,
    required this.isLive,
    required this.pickup,
    required this.dropoff,
    this.courierLat,
    this.courierLng,
    this.etaMinutes,
    this.courierName = '',
    this.courierContact = '',
    this.updatedAt,
    this.proofOfDelivery = '',
    this.deliveredAt,
    this.deliveryNote = '',
  });

  final String orderNumber;
  final String status;
  final String productName;

  /// Whether a courier position is being shared right now.
  final bool isLive;

  final DeliveryPoint pickup;
  final DeliveryPoint dropoff;

  final double? courierLat;
  final double? courierLng;
  final int? etaMinutes;
  final String courierName;
  final String courierContact;
  final DateTime? updatedAt;

  /// The photo the rider took at the door, behind the authenticated files
  /// route. The buyer's own gate is in it, so it is readable only by the three
  /// people on the delivery.
  final String proofOfDelivery;
  final DateTime? deliveredAt;
  final String deliveryNote;

  bool get hasProof => proofOfDelivery.isNotEmpty;

  bool get hasCourier =>
      isLive && courierLat != null && courierLng != null;

  /// Whether there is enough to draw anything at all.
  bool get hasMap => pickup.hasPoint || dropoff.hasPoint || hasCourier;

  /// Tracking only means something once the rice has left the shop.
  bool get isOnTheWay => status == 'shipped' || status == 'delivered';

  factory DeliveryTracking.fromJson(Map<String, dynamic> json) {
    return DeliveryTracking(
      orderNumber: (json['orderNumber'] ?? '').toString(),
      status: (json['status'] ?? '').toString(),
      productName: (json['productName'] ?? '').toString(),
      isLive: json['isLive'] == true,
      pickup: DeliveryPoint.fromJson(json['pickup'] as Map<String, dynamic>?),
      dropoff: DeliveryPoint.fromJson(json['dropoff'] as Map<String, dynamic>?),
      courierLat: (json['lat'] as num?)?.toDouble(),
      courierLng: (json['lng'] as num?)?.toDouble(),
      etaMinutes: (json['etaMinutes'] as num?)?.toInt(),
      courierName: (json['courierName'] ?? '').toString(),
      courierContact: (json['courierContact'] ?? '').toString(),
      updatedAt: DateTime.tryParse((json['updatedAt'] ?? '').toString()),
      proofOfDelivery: (json['proofOfDelivery'] ?? '').toString(),
      deliveredAt: DateTime.tryParse((json['deliveredAt'] ?? '').toString()),
      deliveryNote: (json['deliveryNote'] ?? '').toString(),
    );
  }
}

class DeliveryService {
  DeliveryService._();

  static final DeliveryService instance = DeliveryService._();

  final _api = ApiClient.instance;

  /// Tracking for one order row - one product, from one seller.
  Future<DeliveryTracking> track(String orderId) async {
    final body = await _api.get('/delivery/$orderId');

    if (body is! Map<String, dynamic>) {
      throw ApiException('The server sent no tracking for this order.');
    }

    return DeliveryTracking.fromJson(body);
  }
}
